internal/fxreflect/fxreflect.go

```diff
@@ -25,25 +25,84 @@ import (
 	"reflect"
 	"runtime"
 	"strings"
+
+	"go.uber.org/dig"
 )
 
-// ReturnTypes takes a func and returns a slice of string'd types
-// TODO instead of duplicating the dig's reflect logic, trying to
-// determine which types actually made it into the container, have
-// dig return a Result struct which could contain the ProvidedTypes
+// ReturnTypes takes a func and returns a slice of string'd types.
 func ReturnTypes(t interface{}) []string {
+	// TODO(grayson): instead of duplicating the dig's reflect logic, trying to
+	// determine which types actually made it into the container, have
+	// dig return a Result struct which could contain the ProvidedTypes
+
+	if reflect.TypeOf(t).Kind() != reflect.Func {
+		// Invalid provide, will be logged as an error.
+		return []string{}
+	}
+
 	rtypes := []string{}
-	fn := reflect.ValueOf(t).Type()
+	ft := reflect.ValueOf(t).Type()
 
-	for i := 0; i < fn.NumOut(); i++ {
-		if !isErr(fn.Out(i)) {
-			rtypes = append(rtypes, fn.Out(i).String())
-		}
+	for i := 0; i < ft.NumOut(); i++ {
+		t := ft.Out(i)
+
+		traverseOuts(key{t: t}, func(s string) {
+			rtypes = append(rtypes, s)
+		})
 	}
 
 	return rtypes
 }
 
+// this type is basically straight out of dig, which is a strong signal
+// that exporting it could really DRY up some things for fx-dig relationship.
```

Comment:
Can we do the opposite, exposing less instead of more?

Instead of having Fx get into the reflection game at all, we could create an option to pass a logger to dig containers. Then dig can handle logging, just like it'll eventually need to handle more structured introspection. (Of course, to dig a logger is just something that satisfies a simple interface, like `github.com/uber-go/fx/internal/fxlog.Printer`.)

Dig would also, of course, need to also accept a stack frame identifying the provider or invoker.